File: 223.矩形面积.py
#
# @lc app=leetcode.cn id=223 lang=python3
#
# [223] 矩形面积
#

# @lc code=start
class Solution:
    def computeArea(self, A: int, B: int, C: int, D: int, E: int, F: int, G: int, H: int) -> int:
        def area(A: int, B: int, C: int, D: int):
            return (D-B)*(C-A)

        overlap = [A if A > E else E, B if B > F else F,
                   C if C < G else G, D if D < H else H]
        if overlap[0] >= overlap[2] or overlap[1] >= overlap[3]:
            overlap = [0, 0, 0, 0]
        result = area(A, B, C, D)+area(E, F, G, H)-area(*overlap)
        return result
# @lc code=end

File: 223.矩形面积.cpp
/*
 * @lc app=leetcode.cn id=223 lang=cpp
 *
 * [223] 矩形面积
 */

// @lc code=start
class Solution
{
public:
    int computeArea(int A, int B, int C, int D, int E, int F, int G, int H)
    {
        long width = max((long)0, (long)min(G, C) - max(A, E));
        long height = max((long)0, (long)min(D, H) - max(B, F));
        return (int)((D - B) * (long)(C - A) + (H - F) * (G - E) - width * height);
    }
};
// @lc code=end
